pkg/snet/stun_handler.go

```diff
@@ -0,0 +1,420 @@
+}
+
+type bufferedPacket struct {
+	data []byte
+	addr net.Addr
+}
+
+type stunResponse struct {
+	mappedAddr netip.AddrPort
+	err        error
+}
+
+func newSTUNHandler(conn *net.UDPConn) (*stunHandler, error) {
+	// Get the receive buffer size
+	fd, err := conn.File()
+	if err != nil {
+		return nil, err
+	}
+	defer fd.Close()
+	rcvBufSize, err := syscall.GetsockoptInt(int(fd.Fd()), syscall.SOL_SOCKET, syscall.SO_RCVBUF)
+	if err != nil {
+		return nil, err
+	}
+	maxNumPacket := rcvBufSize / 64 // assuming lower bound of per packet metadata of 64 bytes
+	if maxNumPacket < 10 {
+		maxNumPacket = 10
+	}
+
+	handler := &stunHandler{
+		UDPConn:              conn,
+		recvChan:             make(chan bufferedPacket, maxNumPacket),
+		maxQueuedBytes:       int64(rcvBufSize),
+		recvStunChan:         make(chan []byte, 100),
+		stunChans:            make(map[stun.TxID]chan stunResponse),
+		mappings:             make(map[*net.UDPAddr]*natMapping),
+		retransmissionTimers: make(map[*net.UDPAddr]*retransmissionTimer),
+		pendingRequests:      make(map[*net.UDPAddr]bool),
+	}
+	handler.cond = sync.NewCond(&handler.mutex)
+	return handler, nil
+}
+
+func (c *stunHandler) queuePacket(pkt bufferedPacket) bool {
+	pktLen := int64(len(pkt.data))
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.queuedBytes+pktLen > c.maxQueuedBytes {
+		return false
+	}
+
+	select {
+	case c.recvChan <- pkt:
+		c.queuedBytes += pktLen
+		return true
+	default:
+		return false
+	}
+}
+
+func (c *stunHandler) dequeuePacket() (bufferedPacket, bool) {
+	select {
+	case pkt := <-c.recvChan:
+		c.mutex.Lock()
+		c.queuedBytes -= int64(len(pkt.data))
+		c.mutex.Unlock()
+		return pkt, true
+	default:
+		return bufferedPacket{}, false
+	}
+}
+
+func (c *stunHandler) ReadFrom(b []byte) (int, net.Addr, error) {
+	for {
+		pkt, ok := c.dequeuePacket()
+		if ok {
+			copy(b, pkt.data)
+			return len(pkt.data), pkt.addr, nil
+		}
+		n, addr, err := c.UDPConn.ReadFrom(b)
+		if err != nil {
+			return n, addr, err
+		}
+		if stun.Is(b) {
+			select {
+			case c.recvStunChan <- b[:n]:
+			default:
+			}
+		} else {
+			return n, addr, nil
+		}
+	}
+}
+
+func (c *stunHandler) readStunPacket(ctx context.Context) ([]byte, error) {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+			select {
+			case pkt := <-c.recvStunChan:
+				return pkt, nil
+			default:
+				buf := make([]byte, 1500)
+				n, addr, err := c.UDPConn.ReadFrom(buf)
+				if err != nil {
+					return nil, err
+				}
+				buf = buf[:n]
+				if stun.Is(buf) {
+					return buf, nil
+				} else {
+					c.queuePacket(bufferedPacket{data: buf[:n], addr: addr})
+				}
+			}
+		}
+	}
+}
+
+func (c *stunHandler) getStunResponse(ctx context.Context, txid stun.TxID) (netip.AddrPort, error) {
+	c.mutex.Lock()
+	ch, ok := c.stunChans[txid]
+	c.mutex.Unlock()
+	if !ok {
+		return netip.AddrPort{}, fmt.Errorf("STUN response channel not found")
```

Comment:
Globally pre-allocate this error, find ` = serrors.New(` for examples on how to do this.